[WIP] Unique location ids

vgeorge opened the pull request:
This is a work in progress PR for following up the implementation of unique id for locations.

The current approach is to provide a task, to be run regularly or on-demand, that fetchs unique locations from measurements using parameters `country`, `lon`, `lat`.

Each time this query is run, it performs a upsert on the locations table, inserting new locations or updating metadata (measurement count, parameters, etc...) of existing locations.

At the time of writing, the task can be run on-demand at route `/admin/update-locations?adminKey=adminKey` and will insert new locations at locations table. Parts to be implemented:

- Run query automatically in regular intervals;
- Update existing locations on query run;
- Refactor `/locations` endpoint to fetch data from `locations` table;
- Refactor aggregation endpoints.

jflasher commented on api/services/locations.js:
```diff
@@ -0,0 +1,170 @@
+} = config.get('athena');
+
+// Athena client
+const athena = new AWS.Athena({
```
Is there anyway to reuse the code at https://github.com/openaq/openaq-api/blob/develop/api/services/athena.js or `little-owl` instead of rewriting all new logic here?

jflasher commented on api/services/locations.js:
```diff
@@ -0,0 +1,170 @@
+    .promise();
+
+  if (has(queryResults, 'ResultSet.Rows')) {
+    let rows = queryResults.ResultSet.Rows;
```
Sorry if I am missing the logic to do this, but do you know if this will get **all** the results or does it need to be looped or paged to get all the results?

jflasher commented on api/services/locations.js:
```diff
@@ -0,0 +1,170 @@
+
+// The locations query
+const QueryString = `
+  SELECT
```
What's going to happen to locations with no coordinates? Are we accounting for them here somehow?

jflasher commented on api/services/locations.js:
```diff
@@ -0,0 +1,170 @@
+      return acc;
+    }, {});
+
+    for (const location of locations) {
```
This is being done on locations right? We've currently got 10,353 so this would loop and make 10,353 database requests every time this function was run, right? That sounds bad, but I am willing to be told otherwise. Would it make more sense to pull the whole table into memory once and then compare against that?

jflasher commented on migrations/20190329111919_add-locations.js:
```diff
@@ -0,0 +1,19 @@
+exports.up = function (knex) {
```
Can you point me to where we're getting all the info to go into this table, like `city`, `location`, etc? From the code, I only saw data coming from the Athena query which isn't getting this data, unless I missed it.

vgeorge commented:
I've updated the "locations update" task to address feedback provided. Now it includes metadata about parameters, location names and city into locations table.

@jflasher, you can review this code before I finish refactoring /locations endpoint, if you prefer and have time. A good entry point for reviewing the logic of the update process is the test file `locations-update.js`. 

If you want to test the endpoint to start a location update on the development server, init the machine as described in the README and access:

http://localhost:3004/admin/update-locations?adminToken=adminToken

The process takes about 2 minutes and credentials for Athena must be available. I suggest setting up a local.json file at config dir with AWS credentials, as it is .gitignored.

vgeorge commented:
Added the refactor of `/locations` endpoint to this PR. It will now use data obtained via `locations-update` task. 

There is one breaking change: `nearest` parameter was removed. The documentation was already showing it as **deprecated** because the same effect might be accomplished by using `order_by=distance|limit=10`. Adding support in the new query logic is not trivial, so I've set the endpoint to return "BadRequest" status when nearest is present.

Almost all features (filtering, sorting, geoquery) are working as before. There is a issue I couldn't solve yet, which is search by similar location. Location names are stored into arrays of strings and there is no direct way to search inside them in PostgreSQL. I'll keep looking into it.

jflasher commented on api/routes/admin.js:
```diff
@@ -0,0 +1,41 @@
+'use strict';
```
Can you move this into webhooks.js? It feels like it's serving a very similar purpose of allowing us to trigger the system somehow so I'd like to keep these functional items together.

vgeorge replied:
Yes, makes sense to have this code there.

jflasher commented on api/routes/locations.js:
```diff
@@ -1,21 +1,27 @@
 'use strict';
```
Just sticking it up here since I'm not sure where else it should go.

Looking over the new `/locations` output versus the current, the below items come up
- [ ] `coordinates` should be an object, it's showing up as the PostGIS string. Just noticed there is a lat/lon field added, but that can instead be put in the coordinates field.
- [ ] We are missing `sourceName` which we should keep. It should be the last used sourceName, which should hopefully be the active source.
- [ ] In the new output, there is a NULL `name` field, I think that can be removed?
- [ ] For things like `city` and `location` which are now being treated like arrays, I think that will break a lot of things downstream. Can we change those to `cities` and `locations` as arrays, and keep `city` and `location` to be strings, I guess using the last value that was recorded for them?

vgeorge replied:
Will fix coordinates and implement changes as proposed.

jflasher commented on lib/athena-queries.js:
```diff
@@ -0,0 +1,49 @@
+const { fetchesTable } = config.get('athena');
+
+module.exports = {
+  locationsMetadata: `
```
Does this mean we can get rid of the existing locations query? I think yes right? 

https://github.com/openaq/openaq-api/blob/develop/api/controllers/webhooks.js#L87

Actually, I think this webhook endpoint would crash currently if called since `locations.js` not longer exists.

vgeorge replied:
Didn't noticed that webhooks would broke. I'll review that part so we can have locations Athena sync logic is use there, as propose above.

vgeorge commented on api/services/locations-update.js:
```diff
@@ -0,0 +1,250 @@
+        .update(existingLocation)
+        .where({ id: existingLocation.id });
+    }
+  });
```
Yes, insert operation will be big at the first, then update operation will be the larger one. The update strategy was made that way because current PostgreSQL version (9.4) doesn't support [ON CONFLICT UPDATE](https://www.postgresql.org/docs/9.6/sql-insert.html) operator, which would allow batch upsert. Do you think we can migrate to PostgreSQL 9.6? The impact of moving to 10.x or 11.x it's unclear to me, I wouldn't go for a major release update. Version 9.6 [will be supported until 2021](https://www.postgresql.org/support/versioning/).

vgeorge commented on api/routes/locations.js:
```diff
@@ -1,21 +1,27 @@
  * @apiParam {string=pm25, pm10, so2, no2, o3, co, bc} [parameter] Limit to certain one or more parameters (ex. `parameter=pm25` or `parameter[]=co&parameter[]=pm25`)
  * @apiParam {boolean} [has_geo] Filter out items that have or do not have geographic information.
- * @apiParam {string} [coordinates] Center point (`lat, lon`) used to get locations within/near a certain area. (ex. `coordinates=40.23,34.17`)
- * @apiParam {number} [nearest] Get the X nearest locations to coordinates, must be used with `coordinates`. Wins over `radius` if both are present. Will add `distance` property to locations. DEPRECATED: Use `order_by=distance` with `limit=X` instead.
```
Cool, opened an issue about this on the website repo: https://github.com/openaq/openaq.org/issues/286

vgeorge commented on config/default.json:
```diff
@@ -1,4 +1,9 @@
+  "athena": {
+    "accessKeyId": "",
+    "secretAccessKey": "",
+    "fetchesTable": "fetches.fetches_realtime"
```
Noted, will change.

vgeorge commented:
@jflasher thanks for the feedback. I'm working on requested changes and will get back to you soon. Regarding the rebase, I see that this PR is already based on the latest commit of develop:

![Screenshot 2019-05-06 12 29 57](https://user-images.githubusercontent.com/329694/57222592-0d2c6b00-6ffb-11e9-911e-b50e015da86f.png)

vgeorge commented:
@jflasher, I've finished adding changes to this PR in order to address your last feedback. I've also added Athena auto-sync and updated the README with new configuration options. I believe the refactor to enable unique location ids is done. Can you please review?

jflasher commented on api/routes/cities.js:
```diff
@@ -1,8 +1,13 @@
 'use strict';
```
```json
{
country: "AD",
count: 65220,
locations: 2
}
```
Looks like we're missing `city` and `name` if you want to move people towards that as well.

jflasher commented on api/routes/measurements.js:
```diff
@@ -96,7 +96,6 @@ module.exports = [
       if (request.query) {
```
Ok, I think the last wrinkle I see here (famous last words!). You can see in the old webbooks.js, we generated the total count of all measurements based on what was being returned from the countries aggregation ([here](https://github.com/openaq/openaq-api/blob/develop/api/controllers/webhooks.js#L121)). Now we don't get that and we don't save it to Redis. There is some funky language to handle checking Redis for this value or querying the database to get the count [here](https://github.com/openaq/openaq-api/blob/develop/api/controllers/measurements.js#L35). Given the new approach, I think we need to save the total count to the database and grab it from there. At a minimum, we need to somehow have the total count easy to get (or else all `/measurements` calls will timeout). It'd be better if we could get things like `/measurements?country=CN`, etc also to have the counts pre-calculated (which I know we have in the database now), but that's a bonus.

jflasher replied:
Thanks @vgeorge. I think the way you implemented the counts query will speed up a number of the requests, which is great! 

Please see http://localhost:3004/v1/measurements?value_from=10 as a failing example though. There are a number of query variables supported at https://docs.openaq.org/#api-Measurements that the `locations` table doesn't support. I think it may be best to check to see if the variables passed in the query are all supported by the `locations` table and if not, fall back to querying across the database. I think that code was already in the controller, so hopefully this is just an if/else.

vgeorge replied:
Hey, @jflasher, I've changed the controller, including tests covering the parameters you mentioned.

Now, by default, it will fetch counts from the measurements aggregations available at `locations` table. If `value_*/date_*` is passed, it will fetch from `measurements`. I know this incurs in returning smaller (and incorrect) counts for the second case, but I didn't saw a way of return valid counts for that case if the full history is unavailable. Currently API do not seems to be working when these parameters are used:

https://api.openaq.org/v1/measurements?parameter=co&value_from=100&value_from=1000

jflasher replied:
Thanks @vgeorge, I know this is a bit weird, but I think this is the best we can do right now given how we have the data separated about between a database and S3. If we could just stop using these counts, it would actually make things a lot easier. I blame this on @scisco as I was just following his design on a previous project. :)

I'll try and take a look at this soon and I should hopefully be able to deploy to staging.

scisco replied:
Blame everything on me :)

vgeorge commented:
Hey @jflasher, I've pushed changes to address your latest feedback, please review. Regarding the counts, the approach was to change the `/measurements` controller by making count queries at the `locations` table, instead of Redis. Passing parameters works, so it's possible to filter by country, city and other parameters also available on `/locations` endpoint.

My latest commit didn't triggered a build in TravisCI, do you know why?

vgeorge commented:
Never mind, I could trigger a TravisCI build manually.

jflasher commented:
Looks good to go, working on getting this into staging now!

caparker reviewed (approved):
Looks good. Only thing to point out would be
* trends.py is not updated but we are not using those endpoints currently so I would only update them if we are going to reinstate the trends.
* No tests but we would basically be testing if the pydantic method le=... works so its a gray area.
